<h1>
GenomeVarAPI User Guide
</h1>
<p>Technical Documentation is best viewed from GitHub. See
<a href='https://github.com/ms2206/GenomeVarAPI/tree/main'>https://github.com/ms2206/GenomeVarAPI/tree/main</a>.</p>
<h2>
Overview
</h2>
<div id="overview-section">
<p>GenomeVarAPI is a linux-based tool for parsing and interacting with
VCF (Variant Call Format) files. The tool has 3 elements: a python
parsing module, an sqlite database, and a REST API for end-user
interface.</p>
<p><em> The tool is not guaranteed to work on Windows PC and makes no
claims to be platform independent. </em></p>
<p><img src='./figures/overview.svg' alt='Flowchart of overview'></p>
</div>
<h2>
Getting Started
</h2>
<div id="getting-started-section">
<p>Tool Dependencies: Python, Node.js, npm. Please install these
dependencies before running the tool. For assistance refer to:
<a href='./server_technical_docs.md'>API Technical
Documentation</a>.</p>
<h3>
Download the tool
</h3>
<p>GitHub clone:</p>
<pre><code>git clone https://github.com/ms2206/GenomeVarAPI.git</code></pre>
<p>Change directory: <br></p>
<pre><code>cd GenomeVarAPI</code></pre>
<p>Run the entrypoint.sh: <br></p>
<pre><code>Usage: ./src/entrypoint.sh [--server-only]</code></pre>
<p>Description: This script acts as a wrapper to start the application.
The script checks if the required python environment exists, if not, it
creates the environment. It then initializes the SQLite database, parses
VCF files, installs the node dependencies and starts the Node.js server.
<br></p>
<p>The <code>–server-only</code> option is useful if you just want to
load the server without re-populating the database. This option assumes
node the node packages are already installed. <br></p>
<h3>
Adding VCF files
</h3>
<p>Any VCF files placed in the <code>./data.raw</code> directory will be
parsed by the tool upon initialization. The current release supports
parsing VCF files annotated using SnpEff legacy ‘EFF’ format. The ‘ANN’
is not yet supported but will be made available in a future release.
Considering gene annotations, the tool is limited to extracting gene
names for ‘mRNA’ labeled genes, a future release will expand on these
capabilities. <br><em>Note: if your annotations do not follow this
format the tool should still complete but their annotations and gene
names will not be entered into the database</em><br><br></p>
<p>Flowchart for entrypoint.sh: <br>
<img src='./figures/entrypoint.svg' alt='Flowchart of entrypoint.sh'></p>
<p><i>TODO: Add docker image. use ./src/entrypoint.sh as a
wrapper.</i></p>
</div>
<h2>
Interacting with endpoints
</h2>
<p>Upon running the entrypoint script you shall have a server running on
http://localhost:3000. <br><br> The API has 8 main endpoints:</p>
<div id="api-endpoints-list">
<ul>
<li>
<code>api/genomes</code>
</li>
<li>
<code>api/genomes/:genome_id</code>
</li>
<li>
<code>api/genomes/:genome_id/variants</code>
</li>
<li>
<code>api/genomes/:genome_id/snps</code>
</li>
<li>
<code>api/genomes/:genome_id/indels</code>
</li>
<li>
<code>api/genomes/:genome_id/:chromosome_id/:mbp?</code>
</li>
<li>
<code>api/variants/gene/:gene_name</code>
</li>
<li>
<code>api/variants/:genome_id</code>
</li>
</ul>
<p><br></p>
</div>
<div id="api-endpoints-examples">
<ol>
<li>
List all VCF in database.
</li>
<p><code>http://localhost:3000/api/genomes</code><br><br></p>
<li>
List metadata from a given VCF in database.
</li>
<p><code>http://localhost:3000/api/genomes/RF_041/</code><br><br></p>
<li>
List number of variants in each VCF {genome_id} by chromosome.
</li>
<p><code>http://localhost:3000/api/genomes/RF_001/variants</code><br><br></p>
<li>
List number of SNPs in each VCF {genome_id} by chromosome.
</li>
<p><code>http://localhost:3000/api/genomes/RF_001/snps</code><br><br></p>
<li>
List number of INDELs in each VCF {genome_id} by chromosome.
</li>
<p><code>http://localhost:3000/api/genomes/RF_001/indels</code><br><br></p>
<li>
List genes impacted by moderate or high impact variants in a specific
chromosome region for a specific VCF {genome_id}.
</li>
<p><em>:mbp is an optional parameter to be the number of mbp from the
start of the chromosome. If NULL then the whole chromosome will be
shown.</em><br>
<code>http://localhost:3000/api/genomes/RF_001/chr03/20</code><br><br></p>
<li>
List genomes which contains a variant that impacts a specific gene.
</li>
<p><code>http://localhost:3000/api/variants/gene/Solyc03g006480.1.1</code><br><br></p>
<li>
List variants from a specific genome.
</li>
<code>http://localhost:3000/api/variants/RF_041</code><br><br>
</ol>
</div>
</div>
<h2>
BETA: Plumber API
</h2>
<p>A beta version of an API is available for users to visualize data
from their VCF files. This feature is currently in beta release and has
been intentionally kept separate from the main API hosted at
http://localhost:3000. This decision was made to allow for thorough
testing and to ensure the stability of the Plumber API. Once testing is
complete, the Plumber API will be integrated into the main server. In
the meantime, a separate wrapper can be used to initialize this Plumber
server.</p>
<h3>
Dependencies:
</h3>
<p>R</p>
<h3>
Set Up:
</h3>
<p>To start this server:<br> <code>Rscript src/utils/wrapper_plummer.R
src/api/plumber_API.R</code><br></p>
<p>Navigate to the Swagger API:<br>
<code>http://127.0.0.1:3001/__docs/__</code></p>
<p>OR</p>
<p>Navigate to the local API:<br> <code>Running plumber API at
http://127.0.0.1:3001</code></p>
<h3>
Plumber Endpoints:
</h3>
<ol>
<li>
Barplot of count of variant per genome.
</li>
<p><code>http://127.0.0.1:3001/count_variants?genome_id=&lt;GENOME_ID&gt;</code></p>
<p><img src='./figures/count_var.png' alt='count_var boxplot'></p>
</ol>
<h2>
Documentation
</h2>
<ul>
<li>
<a href='./user_guide.md'>User Documentation</a>
</li>
<li>
<a href='./database_technical_docs.md'>Database Technical
Documentation</a>
</li>
<li>
<a href='./parse_vcf_technical_docs.md'>VCF Parser Technical
Documentation</a>
</li>
<li>
<a href='./server_technical_docs.md'>API Technical Documentation</a>
</li>
</ul>
<h3>
Support and Feedback
</h3>
<p><b>Support Email</b>:<br> For general inquiries and technical support
contact: <em>matspriggs@gmail.com</em><br><br></p>
<p><b>Social Media</b>:<br>
<a href='https://www.linkedin.com/in/matthew-spriggs-324215121/'>LinkedIn</a><br></p>
